### PR TITLE
[GEARPUMP-262] Add setup and teardown to user defined functions

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [x] Make sure tests pass via `sbt clean test`.
 - [x] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

File: examples/streaming/kafka/src/main/scala/org/apache/gearpump/streaming/examples/kafka/dsl/KafkaReadWrite.scala [hunk 1]
```diff
@@ -21,8 +21,8 @@ package org.apache.gearpump.streaming.examples.kafka.dsl
 import java.util.Properties
 
 import org.apache.gearpump.cluster.client.ClientContext
-import org.apache.gearpump.cluster.main.{CLIOption, ArgumentsParser}
-import org.apache.gearpump.streaming.dsl.StreamApp
+import org.apache.gearpump.cluster.main.{ArgumentsParser, CLIOption}
+import org.apache.gearpump.streaming.dsl.scalaapi.StreamApp
```
Comment (manuzhang) [inline examples/streaming/kafka/src/main/scala/org/apache/gearpump/streaming/examples/kafka/dsl/KafkaReadWrite.scala]: dsl packages are restructured into `dsl.api` for common interfaces, `dsl.scalaapi` for Scala and `dsl.javaapi` for Java.

File: examples/streaming/wordcount-java/src/main/java/org/apache/gearpump/streaming/examples/wordcountjava/dsl/WordCount.java [hunk 2]
```diff
@@ -25,12 +25,17 @@
 import org.apache.gearpump.cluster.client.ClientContext;
 import org.apache.gearpump.streaming.dsl.javaapi.JavaStream;
 import org.apache.gearpump.streaming.dsl.javaapi.JavaStreamApp;
+import org.apache.gearpump.streaming.dsl.api.functions.MapFunction;
```
Comment (manuzhang) [inline examples/streaming/wordcount-java/src/main/java/org/apache/gearpump/streaming/examples/wordcountjava/dsl/WordCount.java]: dsl packages are restructured into `dsl.api` for common interfaces, `dsl.scalaapi` for Scala and `dsl.javaapi` for Java.

File: examples/streaming/wordcount-java/src/main/java/org/apache/gearpump/streaming/examples/wordcountjava/dsl/WordCount.java [hunk 3]
```diff
@@ -46,15 +51,13 @@ public static void main(Config akkaConf, String[] args) throws InterruptedExcept
     JavaStream<String> sentence = app.source(new StringSource("This is a good start, bingo!! bingo!!"),
         1, UserConfig.empty(), "source");
 
-    JavaStream<String> words = sentence.flatMap(s -> Arrays.asList(s.split("\\s+")).iterator(),
-        "flatMap");
+    JavaStream<String> words = sentence.flatMap(new Split(), "flatMap");
 
-    JavaStream<Tuple2<String, Integer>> ones = words.map(s -> new Tuple2<>(s, 1), "map");
+    JavaStream<Tuple2<String, Integer>> ones = words.map(new Ones(), "map");
```
Comment (manuzhang) [inline examples/streaming/wordcount-java/src/main/java/org/apache/gearpump/streaming/examples/wordcountjava/dsl/WordCount.java]: With the addition of `setup` and `teardown` methods, Java 8 lambda functions no longer work.

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/api/functions/FilterFunction.scala [hunk 4]
```diff
@@ -15,17 +15,28 @@
...
+  }
+}
 
 /**
- * Function that map a value of type T to value of type R
+ * Returns true to keep the input and false otherwise.
  *
- * @param <T> Input value type
- * @param <R> Output value type
+ * @tparam T Input value type
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/api/functions/FilterFunction.scala]: typo?

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/scalaapi/Stream.scala [hunk 5]
```diff
@@ -36,55 +38,95 @@ class Stream[T](
...
+  }
+
+  /**
+   * Returns a new stream by applying a [[MapFunction]] to each element.
+   *
+   * @param fn [[MapFunction]]
+   * @return A new stream with type [R]
+   */
+  def map[R](fn: MapFunction[T, R], description: String): Stream[R] = {
+    this.flatMap(FlatMapFunction(fn), description)
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/scalaapi/Stream.scala]: Should it be `map` here?

Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/scalaapi/Stream.scala]: never mind, I got it.

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/api/functions/FilterFunction.scala (hunk 4 above)
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/api/functions/FilterFunction.scala]: this is auto-generated and causes `sbt unidoc` to fail. I'm fixing it.

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/scalaapi/functions/FlatMapFunction.scala [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+import org.apache.gearpump.streaming.dsl.api.functions.{FilterFunction, MapFunction}
+import org.apache.gearpump.streaming.dsl.javaapi.functions.{FlatMapFunction => JFlatMapFunction}
+
+import scala.collection.JavaConverters._
+
+object FlatMapFunction {
+
+  def apply[T, R](fn: JFlatMapFunction[T, R]): FlatMapFunction[T, R] = {
+    new FlatMapFunction[T, R] {
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/scalaapi/functions/FlatMapFunction.scala]: why JFlatMapFunction does not call `setup` and `teardown`?

Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/scalaapi/functions/FlatMapFunction.scala]: good catch

**Comment by manuzhang** [comment]
@huafengw the unidoc failure has been fixed

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/functions/FlatMapFunction.scala [hunk 7]
```diff
@@ -15,17 +15,18 @@
...
+ * This Java version of FlatMapFunction returns a java.util.Iterator.
  *
- * @param <T> Input value type
- * @param <Group> Group Type
+ * @param T Input value type
+ * @param R Output value type
  */
-public interface GroupByFunction<T, Group> extends Serializable {
-  Group apply(T t);
+abstract class FlatMapFunction[T, R] extends SerializableFunction {
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/functions/FlatMapFunction.scala]: Do you think just call it JFlatMapFunction will be less confusing?

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/JavaStream.scala [hunk 8]
```diff
@@ -31,23 +31,23 @@ import org.apache.gearpump.streaming.task.Task
 class JavaStream[T](val stream: Stream[T]) {
 
   /** FlatMap on stream */
-  def flatMap[R](fn: FlatMapFunction[T, R], description: String): JavaStream[R] = {
-    new JavaStream[R](stream.flatMap({ t: T => fn(t).asScala }, description))
+  def flatMap[R](fn: functions.FlatMapFunction[T, R], description: String): JavaStream[R] = {
+    new JavaStream[R](stream.flatMap(FlatMapFunction(fn), "flatMap"))
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/JavaStream.scala]: FlatMapFunction(fn: FlatMapFunction) is kind of weird

Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/JavaStream.scala]: updated

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/functions/FlatMapFunction.scala (hunk 7 above)
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/javaapi/functions/FlatMapFunction.scala]: it's already under the `javaapi` package.

**Comment by huafengw** [comment]
+1

**Pull request closed without merging**